src/core/ext/transport/chttp2/transport/chttp2_transport.cc

```diff
@@ -1725,31 +1729,32 @@ class GracefulGoaway : public grpc_core::RefCounted<GracefulGoaway> {
 
   static void OnPingAckLocked(void* arg, grpc_error_handle /* error */) {
     auto* self = static_cast<GracefulGoaway*>(arg);
-    grpc_timer_cancel(&self->timer_);
+    if (self->timer_handle_ != TaskHandle::kInvalid) {
+      if (self->t_->event_engine->Cancel(self->timer_handle_)) {
+        self->Unref();
+      }
+      self->timer_handle_ = TaskHandle::kInvalid;
+    }
     self->MaybeSendFinalGoawayLocked();
     self->Unref();
   }
 
-  static void OnTimer(void* arg, grpc_error_handle error) {
-    auto* self = static_cast<GracefulGoaway*>(arg);
-    if (!error.ok()) {
-      self->Unref();
-      return;
-    }
-    self->t_->combiner->Run(
-        GRPC_CLOSURE_INIT(&self->on_timer_, OnTimerLocked, self, nullptr),
+  void OnTimer() {
```

Comment:
Additional note: after the timer callback fired (a), the handle is erased from the engine's known handles, but we won't reset `timer_handle_` until inside the combiner (b). If `OnPingAckLocked` happens between (a) and (b), this shouldn't cause issue with `Cancel` cancelling the wrong task because of the aba_token in the handle.